cmd/relayer/main.go

```diff
@@ -0,0 +1,412 @@
+	)
+	if err != nil {
+		return err
+	}
+
+	relayerCommission, err := cliutil.ReadUnsignedDecimalFlag(c, flagRelayerCommission)
+	if err != nil {
+		return err
+	}
+
+	// TODO: do we need to restrict potential commission values? eg. 1%, 1.25%, 1.5%, etc
+	// or should we just require a fixed value for now?
+	v := &validator{
+		ctx:               ctx,
+		ec:                ec,
+		relayerCommission: relayerCommission,
+		forwarderAddress:  forwarderAddr,
+	}
+
+	// the forwarder contract is fixed here; thus it needs to be the same
+	// as what's hardcoded in the swap contract addr for that network.
+	rcfg := &relayer.Config{
+		Ctx:                     ctx,
+		EthClient:               ec,
+		Forwarder:               rcontracts.NewIForwarderWrapped(forwarder),
+		Key:                     key,
+		ValidateTransactionFunc: v.validateTransactionFunc,
+	}
+
+	r, err := relayer.NewRelayer(rcfg)
+	if err != nil {
+		return err
+	}
+
+	if c.Bool(flagWithNetwork) {
+		// cfg.DataDir already has a default set, so only override if the user explicitly set the flag
+		var datadir string
+		if c.IsSet(flagDataDir) {
+			datadir = c.String(flagDataDir) // override the value derived from `flagEnv`
+		} else {
+			datadir = config.DataDir
+		}
+		datadir = path.Join(datadir, "relayer")
+		if err = common.MakeDir(datadir); err != nil {
+			return err
+		}
+
+		h, err := setupNetwork(ctx, c, ec, r, datadir) //nolint:govet
+		if err != nil {
+			return err
+		}
+
+		defer func() {
+			_ = h.Stop()
+		}()
+	}
+
+	if c.Bool(flagRPC) {
+		go signalHandler(ctx, cancel)
+		rpcCfg := &rrpc.Config{
+			Ctx:     ctx,
+			Address: fmt.Sprintf("127.0.0.1:%d", port),
+			Relayer: r,
+		}
+
+		server, err := rrpc.NewServer(rpcCfg) //nolint:govet
+		if err != nil {
+			return err
+		}
+
+		err = server.Start()
+		if errors.Is(err, context.Canceled) || errors.Is(err, http.ErrServerClosed) {
+			return nil
+		}
+	} else {
+		signalHandler(ctx, cancel)
+	}
+
+	return err
+}
+
+func setupNetwork(
+	ctx context.Context,
+	c *cli.Context,
+	ec *ethclient.Client,
+	r *relayer.Relayer,
+	datadir string,
+) (*net.Host, error) {
+	chainID, err := ec.ChainID(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var bootnodes []string
+	if c.IsSet(flagBootnodes) {
+		bootnodes = cliutil.ExpandBootnodes(c.StringSlice(flagBootnodes))
+	}
+
+	listenIP := "0.0.0.0"
+	netCfg := &p2pnet.Config{
+		Ctx:        ctx,
+		DataDir:    datadir,
+		Port:       uint16(c.Uint(flagLibp2pPort)),
+		KeyFile:    c.String(flagLibp2pKey),
+		Bootnodes:  bootnodes,
+		ProtocolID: fmt.Sprintf("/%s/%d/%s", swapnet.ProtocolID, chainID.Int64(), net.ProtocolID),
+		ListenIP:   listenIP,
+	}
+
+	cfg := &net.Config{
+		Context:              ctx,
+		P2pConfig:            netCfg,
+		TransactionSubmitter: r,
+		IsRelayer:            true,
+	}
+
+	h, err := net.NewHost(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	err = h.Start()
+	if err != nil {
+		return nil, err
+	}
+
+	return h, nil
+}
+
+func deployOrGetForwarder(
```

Comment:
Is there another package that this might fit better in? It certainly doesn't look like a function that belongs in `main.go`.

Reply:
there's a `getOrDeploySwapFactory` inside cmd/swapd that's analogous, I can move this to a diff file to separate it?

Reply:
Yeah, a different file from `main.go` would be good.